### ISSUE TITLE
Allow blank lines in list/record literals?

### ISSUE THREAD

**Issue by rtfeldman** [avh4/elm-format#97]
Similarly to how blank lines can separate imports into logical groups, I've run into cases where I want to do the same to long records (e.g. Models) or long list literals (e.g. in DSLs), and currently those blank lines get lost after the elm-format pass.

I don't think there's any more downside to allowing this than there was allowing it in import statements, and it'd make some of my code easier to read.

Thoughts?

**Comment by cpdean** [comment]
What if blank lines were still removed, but comments weren't? 

Currently elm-format turns this
```
type alias Person = {
    name: String,
    age: Int,
    ssn: (Int, Int, Int),

    hobbies: List String,
    favoriteFoods: List String,

    employers: List Int,
    schools: List Int
    }
```

into this
```
module People (..) where


type alias Person =
    { name : String
    , age : Int
    , ssn : ( Int, Int, Int )
    , hobbies : List String
    , favoriteFoods : List String
    , employers : List Int
    , schools : List Int
    }
```

And while that removes the semantic groupings of fields, it would force me to actually document the groupings with comments:
```
type alias Person =
    { -- PII
      name : String
    , age : Int
    , ssn : ( Int, Int, Int )
      -- Recreation
    , hobbies : List String
    , favoriteFoods : List String
      -- Job Stuff
    , employers : List Int
    , schools : List Int
    }

```

But right now elm-format actually removes comments in record definitions:
```
type alias Person =
    { name : String
    , age : Int
    , ssn : ( Int, Int, Int )
    , hobbies : List String
    , favoriteFoods : List String
    , employers : List Int
    , schools : List Int
    }
```

**Comment by avh4** [comment]
Comments in records, tuples and lists will be kept  (see #3 for progress on that).

I think we will probably want to add this before a 1.0 release, but I have been saving it for later because I think we haven't fully explored the question of whether long lists and large records are an anti-pattern that should be discouraged.

**Comment by rtfeldman** [comment]
Ah, gotcha.

Personally I definitely have not found large lists/records to be an anti-pattern. In other languages they've felt unwieldy, but in Elm adding more (use-case necessary) fields to a model while keeping a flat field structure feels totally great. :smile_cat:

**Comment by mgold** [comment]
FYI the second example in @cpdean's comment is now formatted as:
```
type alias Person =
  { -- PII
    name : String
  , age : Int
  , ssn :
      ( Int, Int, Int )
      -- Recreation
  , hobbies : List String
  , favoriteFoods :
      List String
      -- Job Stuff
  , employers : List Int
  , schools : List Int
  }
```
Better, but what's up with the line wrap?

**Comment by avh4** [comment]
After the changes in #162, the following will be possible to group records:
```
type alias Person =
    { -- PII
      name : String
    , age : Int
    , ssn : ( Int, Int, Int )
    , -- Recreation
      hobbies : List String
    , favoriteFoods : List String
    , -- Job Stuff
      employers : List Int
    , schools : List Int
    }
```

and lists:
```
roles =
    [ -- Executives
      "CEO"
    , "CTO"
    , -- Managers
      "Eng Manager 1"
    , "Design Manager 2"
    ]
```

Are blank lines still needed?

**Comment by avh4** [comment]
I think it would be interesting to have something like a comment in the list creates a section.

```elm
roles =
    [ "CEO"
    , "CTO"

    -- Managers
    , "Eng Manager 1"
    , "Design Manager 2"
    ]
```

The problem with this is that there's not a good way to deal with a comment at the beginning of the list (before "CEO").

What if it were like this:

```elm
type alias Person =
  { -- PII
    name : String
  , age : Int
  , ssn : ( Int, Int, Int )

    -- Recreation
  , hobbies : List String
  , favoriteFoods : List String
    
    -- Job Stuff
  , employers : List Int
  , schools : List Int
  }
```

If we did something like that, would it be better to have the section headers indented, or aligned with the commas? Like this:

```elm
type alias Person =
  {
  -- PII
    name : String
  , age : Int
  , ssn : ( Int, Int, Int )

  -- Recreation
  , hobbies : List String
  , favoriteFoods : List String
    
  -- Job Stuff
  , employers : List Int
  , schools : List Int
  }
```

**Comment by avh4** [comment]
Aligning the `--` with the commas appears to be the preferable choice.  It seems like we might just have to live with comments at the very beginning not working very well.

**Comment by strickinato** [comment]
I'd like to call attention again to mgold's comment about keeping the two use cases separate. 

I've had much more use for commenting on an individual fields then breaking them up into sections, 

Currently:

```elm
type alias Thing =
    { horse : String -- Hello
    , dog : Float
    , fish : Integer
    }
```

becomes:

```elm
type alias Thing =
    { horse :
        String
        -- Hello
    , dog : Float
    , fish : Integer
    }
```

**Comment by avh4** [comment]
Trailing comments on record fields is fixed in https://github.com/avh4/elm-format/issues/162.  I am going to delete comments related to that from this thread.

**Comment by avh4** [comment]
Fixed for records in ca1d5daf...4c69c40b
